cmd/sys-mgmt-executor/main.go

```diff
@@ -148,19 +166,99 @@ func runDockerCommands(service string, operation string) error {
 			expectedStatus = true
 		case STOP:
 			expectedStatus = false
+		case METRICS:
+
 		default:
-			// impossible to get here
 			panic(fmt.Sprintf("invalid operation %s", operation))
 		}
 
 		correctStatus, err := checkDockerContainerStatus(service, expectedStatus)
 		if err != nil {
-			return err
+			return "", err
 		}
 		if !correctStatus {
-			return fmt.Errorf("docker %s operation failed for service %s", operation, service)
+			return "", fmt.Errorf("docker %s operation failed for service %s", operation, service)
+		}
+		return "", nil
+	} else if operation == METRICS {
+
+		LoggingClient.Info(fmt.Sprintf("run the docker stats on service (%s) for operation (%s)", service, operation))
+		cmd := exec.Command("docker", "stats")
+		LoggingClient.Info(fmt.Sprintf("finished running the docker stats command"))
+
+		var stdout, stderr []byte
+		var errStdout, errStderr error
+		stdoutIn, _ := cmd.StdoutPipe()
+		stderrIn, _ := cmd.StderrPipe()
+		err := cmd.Start()
+		if err != nil {
+			log.Fatalf("cmd.Start() failed with '%s'\n", err)
+		}
+
+		// Note that cmd.Wait() should be called only AFTER we finish reading
+		// from stdoutIn and stderrIn.
+		// The wait group (wg) ensures that we finish
+		var wg sync.WaitGroup
+		wg.Add(1)
+		ticker := time.NewTicker(time.Second)
+		go func(ticker *time.Ticker) {
+			stdout, errStdout = copyAndCapture(os.Stdout, stdoutIn)
+			wg.Done()
+			now := time.Now()
+			for range ticker.C {
+				LoggingClient.Info(fmt.Sprintf("%s", time.Since(now)))
+			}
+		}(ticker)
+
+		// Create a timer that will kill the process
+		timer := time.NewTimer(time.Second * 3)
+		go func(timer *time.Timer, ticker *time.Ticker, cmd *exec.Cmd) {
+			for range timer.C {
+				cmd.Process.Signal(os.Kill)
+				ticker.Stop()
+			}
+		}(timer, ticker, cmd)
+		stderr, errStderr = copyAndCapture(os.Stderr, stderrIn)
+
+		wg.Wait()
+
+		err = cmd.Wait()
+		if err != nil {
+			LoggingClient.Error(fmt.Sprintf("cmd.Run() failed with %s\n", err.Error()))
+			log.Fatalf("cmd.Run() failed with %s\n", err)
+		}
+		if errStdout != nil || errStderr != nil {
+			log.Fatal("failed to capture stdout or stderr\n")
+			LoggingClient.Error(fmt.Sprintf("failed to capture stdout or stderr\n"))
+		}
+		outStr, errStr = string(stdout), string(stderr)
+		LoggingClient.Info(fmt.Sprintf("\nout:\n%s\nerr:\n%s\n", outStr, errStr))
+		LoggingClient.Info("invocation of metrics executor in runDockerCommands() succeeded...")
+		return outStr, nil
+	} else {
+		return "", fmt.Errorf("system management was requested to perform an unknown operation %s on the service %s", operation, service)
+	}
+}
+
+func copyAndCapture(w io.Writer, r io.Reader) ([]byte, error) {
```

Comment:
Would this not work? https://golang.org/pkg/io/#Copy

Reply:
I compared the signatures of the two functions (i.e. the one you mention, and the one I've used) and notice some differences... The implementation I'm using comes from a Go Cookbook [Advanced command execution in Go with os/exec](https://blog.kowalczyk.info/article/wOYk/advanced-command-execution-in-go-with-osexec.html) and which certainly reflects that thought and care has been taken in putting it together. If possible, I, too, would prefer using a call to the Go standard library. Thanks for mentioning 👍